src/od/od.rs

```diff
@@ -74,418 +96,336 @@ pub fn uumain(args: Vec<String>) -> i32 {
+        });
+    let print_width_line = print_width_block * (line_bytes / byte_size_block);
+
+    if byte_size_block > MAX_BYTES_PER_UNIT {
+        panic!("{}-bits types are unsupported. Current max={}-bits.",
+                8 * byte_size_block,
+                8 * MAX_BYTES_PER_UNIT);
+    }
 
-fn odfunc(input_offset_base: &Radix, fnames: &[InputSource], formats: &[OdFormat]) -> i32 {
+    let mut spaced_formatters: Vec<SpacedFormatterItemInfo> = formats
+        .iter()
+        .map(|f| SpacedFormatterItemInfo { frm: *f, spacing: [0; MAX_BYTES_PER_UNIT] })
+        .collect();
+
+    // calculate proper alignment for each item
+    for sf in &mut spaced_formatters {
+        let mut byte_size = sf.frm.formatter_item_info.byte_size;
+        let mut items_in_block = byte_size_block / byte_size;
+        let thisblock_width = sf.frm.formatter_item_info.print_width * items_in_block;
+        let mut missing_spacing = print_width_block - thisblock_width;
+
+        while items_in_block > 0 {
+            let avg_spacing: usize = missing_spacing / items_in_block;
+            for i in 0..items_in_block {
+                sf.spacing[i * byte_size] += avg_spacing;
+                missing_spacing -= avg_spacing;
+            }
+            // this assumes the size of all types is a power of 2 (1, 2, 4, 8, 16, ...)
+            items_in_block /= 2;
+            byte_size *= 2;
+        }
+    }
 
-    let mut mf = MultifileReader::new(fnames);
-    let mut addr = 0;
-    let bytes = &mut [b'\x00'; LINEBYTES];
     loop {
         // print each line data (or multi-format raster of several lines describing the same data).
+        // TODO: we need to read more data in case a multi-byte sequence starts at the end of the line
 
-        print_with_radix(input_offset_base, addr); // print offset
-		// if printing in multiple formats offset is printed only once
-
-        match mf.f_read(bytes) {
-            Ok(0) => {
-                print!("\n");
+        match input.peek_read(bytes.as_mut_slice(), PEEK_BUFFER_SIZE) {
+            Ok((0, _)) => {
+                print_final_offset(input_offset_base, addr);
                 break;
             }
-            Ok(n) => {
-                let mut first = true; // First line of a multi-format raster.
-                for f in formats {
-                    if !first {
-                        // this takes the space of the file offset on subsequent
-                        // lines of multi-format rasters.
-                        print!("       ");
+            Ok((n, peekbytes)) => {
+                // not enough byte for a whole element, this should only happen on the last line.
+                if n != line_bytes {
+                    // set zero bytes in the part of the buffer that will be used, but is not filled.
+                    let mut max_used = n + MAX_BYTES_PER_UNIT;
```

Comment:
In ```od.rs```, ```parse_formats.rs``` and to a lesser extent ```MultiFileReader.rs``` there are currently some lengthy functions that should be at some point broken up into separate functions so as to make it easier to understand at a glance purpose different sections of their code do / should serve. 

 (e.g. this code could be moved into a small function called ```zero_remaining_buffer_bytes()``` or something like that).